lib/api/2.0/ibms.js

```diff
@@ -0,0 +1,118 @@
+// Copyright 2016, EMC, Inc.
+
+'use strict';
+
+var injector = require('../../../index.js').injector;
+var controller = injector.get('Http.Services.Swagger').controller;
+var schemaApiService = injector.get('Http.Api.Services.Schema');
+var nameSpace = '/api/2.0/ibms/definitions/';
+var waterline = injector.get('Services.Waterline');
+var eventsProtocol = injector.get('Protocol.Events');
+var Errors = injector.get('Errors');
+var _ = injector.get('_'); // jshint ignore:line
+
+
+// GET /api/2.0/ibms
+var ibmsGet = controller(function(req) {
+    return waterline.ibms.find(req.query);
+});
+
+// PUT /api/2.0/ibms
+var ibmsPut = controller({success: 201}, function(req) {
+    var nodeId = req.swagger.params.body.value.nodeId;
+    delete req.swagger.params.body.value.nodeId;
+    return waterline.ibms.upsertByNode(nodeId, req.swagger.params.body.value);
+});
+
+// GET /api/2.0/ibms/identifier
+var ibmsGetById = controller(function(req) {
+    return waterline.ibms.needByIdentifier(req.swagger.params.identifier.value);
+});
+
+// PATCH /api/2.0/ibms/identifier
+var ibmsPatchById = controller( function(req) {
+    var ibmId = req.swagger.params.identifier.value;
+    var values = req.swagger.params.body.value;
+    return waterline.ibms.needByIdentifier(ibmId)
+        .then(function(oldIbm) {
+            /* Get nodes that need to publish events */
+            if (oldIbm.node && !values.nodeId) {
+                return Promise.all([waterline.nodes.getNodeById(oldIbm.node)]);
+            } else if (!oldIbm.node && values.nodeId) {
+                return Promise.all([waterline.nodes.getNodeById(values.node)]);
+            } else if (oldIbm.node && values.nodeId && oldIbm.node === values.nodeId) {
+                return Promise.all([waterline.nodes.getNodeById(oldIbm.node)]);
+            } else if (oldIbm.node && values.nodeId && oldIbm.node !== values.nodeId) {
+                return Promise.all([waterline.nodes.getNodeById(oldIbm.node),
+                        waterline.nodes.getNodeById(values.nodeId)]);
+            }
+        })
+        .then(function(oldNodes) {
+            return waterline.ibms.updateByIdentifier(ibmId, values)
+            .tap(function() {
+                /* Publish events of nodes got beofre update */
+                _.forEach(oldNodes, function(oldNode) {
+                    if (oldNode) {
+                        /* asynchronous, don't wait promise return for performance*/
+                        return waterline.nodes.getNodeById(oldNode.id)
+                        .then(function(newNode) {
+                            return eventsProtocol.publishNodeAttrEvent(oldNode, newNode, 'ibms');
+                        })
+                        .catch(function (error) {
```

Comment:
@srinia6 It's probably worth appending `error.message` to your base error message.